File: src/Modules/Billing/ErpClink.Modules.Billing.Application/Invoices/Models/InvoiceModels.cs
namespace ErpClink.Modules.Billing.Application.Invoices.Models;

public sealed record InvoiceLineInput(
    Guid? ServiceId,
    Guid? PackageId,
    decimal Quantity,
    decimal? LineDiscountAmount,
    int? SortOrder);

public sealed record CreateDraftInvoiceRequest(
    Guid PatientId,
    Guid? MedicalVisitId,
    DateOnly InvoiceDate,
    string CurrencyCode,
    string? Notes,
    decimal? InvoiceDiscountAmount,
    IReadOnlyList<InvoiceLineInput>? Lines);

public sealed record UpdateDraftInvoiceRequest(
    DateOnly InvoiceDate,
    string? Notes,
    decimal? InvoiceDiscountAmount,
    byte[]? RowVersion);

public sealed record AddInvoiceLineRequest(
    InvoiceLineInput Line,
    byte[]? RowVersion);

public sealed record UpdateInvoiceLineRequest(
    decimal Quantity,
    decimal LineDiscountAmount,
    int SortOrder,
    byte[]? RowVersion);

public sealed record VoidInvoiceRequest(string? Reason, byte[]? RowVersion);

public sealed record SearchInvoicesRequest(
    Guid? PatientId,
    Guid? MedicalVisitId,
    string? InvoiceNumber,
    DateOnly? DateFrom,
    DateOnly? DateTo,
    string? Status,
    int Page = 1,
    int PageSize = 20);

public sealed record PatientInvoicesRequest(
    DateOnly? DateFrom,
    DateOnly? DateTo,
    string? Status,
    int Page = 1,
    int PageSize = 20);

public sealed record InvoiceLineDto(
    Guid Id,
    string Source,
    Guid? ServiceId,
    Guid? PackageId,
    Guid? ServicePriceId,
    string DescriptionSnapshot,
    string? ServiceCodeSnapshot,
    decimal Quantity,
    decimal UnitPrice,
    decimal LineDiscountAmount,
    decimal LineSubtotal,
    decimal LineTotal,
    string CurrencyCode,
    int SortOrder);

public sealed record InvoiceDto(
    Guid Id,
    Guid OrganizationId,
    Guid BranchId,
    string? InvoiceNumber,
    Guid PatientId,
    Guid? MedicalVisitId,
    DateOnly InvoiceDate,
    string Status,
    string CurrencyCode,
    decimal SubTotal,
    decimal DiscountAmount,
    decimal TaxAmount,
    decimal TotalAmount,
    decimal PaidAmount,
    decimal OutstandingAmount,
    string? Notes,
    DateTime? IssuedAtUtc,
    string? IssuedBy,
    DateTime? VoidedAtUtc,
    string? VoidedBy,
    string? VoidReason,
    DateTime CreatedAtUtc,
    string? CreatedBy,
    DateTime? UpdatedAtUtc,
    string? UpdatedBy,
    byte[] RowVersion,
    IReadOnlyList<InvoiceLineDto> Lines);

public sealed record PagedInvoicesResult(
    IReadOnlyList<InvoiceDto> Items,
    int TotalCount,
    int Page,
    int PageSize);

public sealed record PatientOutstandingDto(Guid PatientId, decimal OutstandingAmount, string CurrencyCode);

File: frontend/erpclink-web/src/app/features/billing/models/invoice.models.ts
export interface InvoiceLineInput {
  serviceId?: string | null;
  packageId?: string | null;
  quantity: number;
  lineDiscountAmount?: number | null;
  sortOrder?: number | null;
}

export interface CreateDraftInvoiceRequest {
  patientId: string;
  medicalVisitId?: string | null;
  invoiceDate: string;
  currencyCode: string;
  notes?: string | null;
  invoiceDiscountAmount?: number | null;
  lines?: InvoiceLineInput[] | null;
}

export interface UpdateDraftInvoiceRequest {
  invoiceDate: string;
  notes?: string | null;
  invoiceDiscountAmount?: number | null;
  rowVersion?: string | null;
}

export interface RecordPaymentRequest {
  invoiceId: string;
  paymentDate: string;
  amount: number;
  method: string;
  referenceNumber?: string | null;
  notes?: string | null;
  invoiceRowVersion?: string | null;
}

export interface InvoiceLineDto {
  id: string;
  source: string;
  serviceId?: string | null;
  packageId?: string | null;
  unitPrice: number;
  quantity: number;
  lineTotal: number;
  descriptionSnapshot: string;
}

export interface InvoiceDto {
  id: string;
  invoiceNumber?: string | null;
  patientId: string;
  status: string;
  currencyCode: string;
  subTotal: number;
  discountAmount: number;
  totalAmount: number;
  paidAmount: number;
  outstandingAmount: number;
  notes?: string | null;
  invoiceDate: string;
  rowVersion: string;
  lines: InvoiceLineDto[];
}

export interface PagedInvoicesResult {
  items: InvoiceDto[];
  totalCount: number;
  page: number;
  pageSize: number;
}

export interface PatientOutstandingDto {
  patientId: string;
  outstandingAmount: number;
  currencyCode: string;
}
